runtime/src/read_only_accounts_cache.rs

```diff
@@ -92,86 +140,86 @@ impl ReadOnlyAccountsCache {
-
-            if try_remove {
-                self.cache.remove(&key);
-            }
-            current_size = current_size.saturating_sub(*size); // we don't subtract on remove, so subtract now
-            if current_size <= self.max_data_size {
-                break;
-            }
         }
-        lru.drain(0..processed);
-        current_size
+        if freed_bytes > 0 {
+            // if this overflows, we'll have a really big data size, so we'll clean everything, scan all, and reset the size. Not ideal, but not terrible.
+            self.data_size.fetch_sub(freed_bytes, Ordering::Relaxed);
+        }
+        done
     }
 
-    fn calculate_lru_list(&self, lru: &mut Vec<LruEntry>) -> usize {
-        // purge in lru order
+    fn calculate_lru_list(&self) -> (Vec<LruEntry>, usize) {
+        let mut lru = Vec::with_capacity(self.cache.len());
         let mut new_size = 0;
         for item in self.cache.iter() {
             let value = item.value();
-            let item_len = value.account.data().len() + self.per_account_size;
+            let item_len = self.account_size(&value.account);
             new_size += item_len;
-            lru.push((*value.last_used.read().unwrap(), item_len, *item.key()));
+            lru.push((*value.last_used.read().unwrap(), *item.key()));
         }
-        new_size
+        (lru, new_size)
     }
 
-    fn maybe_purge_lru_items(&self, new_item_len: usize) -> usize {
-        let mut new_size = *self.data_size.read().unwrap() + new_item_len;
-        if new_size <= self.max_data_size {
-            return new_size;
-        }
+    fn bg_purge_lru_items(&self, once: bool) {
+        let mut lru = Vec::new();
+        let mut lru_index = 0;
+        let mut stop = false;
+        loop {
+            if !once {
+                sleep(Duration::from_millis(200));
+            } else {
+                if stop {
+                    break;
+                }
+                stop = true;
+            }
 
-        // purge from the lru list we last made
-        let mut list = self.lru.write().unwrap();
-        new_size = self.purge_lru_list(&mut list, true, new_size);
-        if new_size <= self.max_data_size {
-            return new_size;
-        }
+            if self.stop.load(Ordering::Relaxed) {
+                break;
+            }
 
-        // we didn't get enough, so calculate a new list and keep purging
-        new_size = self.calculate_lru_list(&mut list);
-        if new_size > self.max_data_size {
-            list.sort();
-            new_size = self.purge_lru_list(&mut list, false, new_size);
-            // the list is stored in self so we use it to purge next time
+            // purge from the lru list we last made
+            if self.purge_lru_list(&lru, &mut lru_index) {
+                continue;
+            }
+
+            // we didn't get enough, so calculate a new list and keep purging
+            let (new_lru, new_size) = self.calculate_lru_list();
```

Comment:
What about `lru.clear()` here and then just recreate into it again?

Reply:
done. bench tps shows a range of 1.3-1.8M entries in this list. With this new strategy, this lru list will keep growing and then remain at its largest size forever. Not a huge deal. I was imagining there were advantages to it using exactly what it needed each time the lru list is recalculated, which isn't too often. I can also add code to reallocate it smaller if we shrink size enough. I was trying to keep the code simple with less things to go wrong.